### PR TITLE
Added calcBestShotOnFriendlyGoal functions and tests

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Added `calcBestShotOnFriendlyGoal` evaluation functions and restructured the `calcBestShot` functions a bit to more easily support both teams. Also removed the `calcBestShotAll` versions because YAGNI.

### Testing Done

Unit tests added

### Resolved Issues

None

### Length Justification

A fair amount of stuff was changed or moved so it adds up to just over 500 lines

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.h [hunk 1]
```diff
@@ -9,82 +8,133 @@
...
-     * length and goal width
-     *
-     * @param obstacles-obstacles is a list of all the obstacles in the way between the
-     * point and the net
+     * Finds the best shot on the given goal, and returns the best target to shoot at and
+     * the largest open angle interval for the shot (this is the total angle between the
+     * obstacles on either side of the shot vector).
+     *
+     * @param goal_post_neg The goal post of the net with the negative y-coordinate
+     * @param goal_post_posThe goal post of the net with the positive y-coordinate
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.h]:
``` suggestion
     * @param goal_post_pos The goal post of the net with the positive y-coordinate
```

File: src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.h [hunk 2]
```diff
@@ -9,82 +8,133 @@
...
+     * be treated as a shot at the friendly goal
+     *
+     * @return the best target to shoot at and the largest open angle interval for the
+     * shot (this is the total angle between the obstacles on either side of the shot
+     * vector).
+     */
+    std::pair<Point, Angle> calcBestShotOnGoal(const World &world, const Point &point,
+                                               double radius,
+                                               const std::vector<Robot> &robots_to_ignore,
+                                               bool shoot_at_enemy_goal);
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.h:55]: I'm a bit confused, the `shoot_at_enemy_goal` doesn't appear anywhere else in this PR? Regardless, this seems a kinda icky way to do this, I think just have the two separate functions is probably a better call.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking pretty good to me, no major comments. I should be able to make this fit in with my changes fairly neatly.

**Review by EvanMorcom — COMMENTED** [review]
Looks pretty good to me.

File: src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.cpp [hunk 3]
```diff
@@ -4,79 +4,102 @@
...
+        std::pair<Point, Angle> best_shot;
+
+        // Calculate the best_shot based on what goal we're shooting at
+        if (shoot_on_enemy_goal)
         {
-            obstacles.push_back(i.position());
+            best_shot = calcBestShotOnGoal(world.field().enemyGoalpostNeg(),
+                                           world.field().enemyGoalpostPos(), point,
+                                           obstacles, radius);
+            // If not shot is found, at least set the target to the goal
```
[the remark this answers is not in the document]
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.cpp]: fixed

File: src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/calc_best_shot.h]: fixed

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**